Added support for dynamic layout renderer in log level filters (e.g. minLevel, maxLevel)

snakefoot opened the pull request:
Extending #2709 and resolves #1502

allowes:

```xml
<nlog>
   <variable name='myLevel' value='debug'/>
    <rules>
      <logger minLevel='${myLevel}'/>
    </rules>
</nlog>
```

And you do this:

```c#
LogManager.KeepVariablesOnReload = true;
LogManager.Configuration.Variables["myLevel"] = "Info";
LogManager.Configuration.Reload();
```

304NotModified commented:
cool!

304NotModified commented on src/NLog/Config/LoggingRuleLevelFilter.cs:
```diff
@@ -0,0 +1,251 @@
+    using NLog.Internal;
+    using NLog.Layouts;
+
+    internal interface ILoggingRuleLevelFilter
```
please move all interfaces and classes to a new file, thanks!

snakefoot replied:
I just repeated how it was done for `LoggerNameMatcher`. Have I done something different?

304NotModified replied:
Yes, I think it was also a comment there. :/

For easier finding, refactoring and consistency, separate files are really preferred. 

If you move them when they are getting to big/complex, you will "lose" history in git.

snakefoot replied:
Done

304NotModified commented on src/NLog/Config/LoggingRuleLevelFilter.cs:
```diff
@@ -0,0 +1,251 @@
+                return this;
+            }
+
+            return new LoggingRuleLevelFilter().SetLoggingLevels(minLevel, maxLevel, enable);
```
this is a bit unexpected. Why the `new LoggingRuleLevelFilter()`. The method is now a combination of "set" and "create"

snakefoot replied:
Because it is difficult to combine dynamic filters with hardcoded assignment of levels. So it converts dynamic-filter to simple-filter if needed. If already simple-filter then it just modifies and returns this.

snakefoot replied:
Have now changed it into `GetSimpleFilterForUpdate()` that returns `LoggingRuleLevelFilter` (that has the method `SetLoggingLevels`)

304NotModified commented on src/NLog/Config/LoggingRuleLevelFilter.cs:
```diff
@@ -0,0 +1,251 @@
+            return new LoggingRuleLevelFilter(LogLevels).SetLoggingLevels(minLevel, maxLevel, enable);
+        }
+
+        private bool[] GenerateLogLevels()
```
please split this method in smaller ones, maybe some code could be deduplicated?

snakefoot replied:
Fixed

304NotModified commented on src/NLog/Config/LoggingRuleLevelFilter.cs:
```diff
@@ -0,0 +1,251 @@
+                LogLevel maxLevel = ParseLogLevel(maxLevelFilter, LogLevel.MaxLevel) ?? LogLevel.MinLevel;
+
+                bool[] logLevels = new bool[LogLevel.MaxLevel.Ordinal + 1];
+                for (int i = minLevel.Ordinal; i <= Math.Min(logLevels.Length - 1, maxLevel.Ordinal); ++i)
```
PS I always use `i<= logLevels.Length - 1 && i < maxLevel.Ordinal` for those cases

snakefoot replied:
Fixed

304NotModified commented on src/NLog/Config/LoggingRuleLevelFilter.cs:
```diff
@@ -0,0 +1,251 @@
+            return activeFilter.Value;
+        }
+
+        LogLevel ParseLogLevel(string logLevel, LogLevel defaultLevel)
```
as there is an argument defaultLevel here, I think it would be logical to have also levelOnError here. Maybe rename defaultLevel to levelIfEmpty.

(defaults are sometimes also used for errors)

snakefoot replied:
Fixed

snakefoot commented on src/NLog/Config/LoggingConfigurationParser.cs:
```diff
@@ -669,31 +667,60 @@ private LoggingRule ParseRuleElement(ILoggingConfigurationElement loggerElement)
+                }
+                else
+                {
+                    if (enableLevels.IndexOf(',') >= 0)
```
Fixed

snakefoot commented:
@304NotModified Not sure what to do now. Have made heavy refactoring of the unit-tests to make them simpler. And now you have committed.

304NotModified commented:
rebase or force push ;)

If you think it's better and rebase is too/unneeded difficult, then force push (and thus remove my commits).

snakefoot commented:
@304NotModified Performed a manual rebase of all your relevant changes, and then a force push.

snakefoot commented on tests/NLog.UnitTests/Config/RuleConfigurationTests.cs:
```diff
@@ -589,5 +589,185 @@ public void LoggingRule_LevelOff_NotSetAsActualLogLevel()
+                if (LogLevel.FromOrdinal(i) == expectedLogLevel)
+                    Assert.True(b.IsEnabled(LogLevel.FromOrdinal(i)));
+                else
+                    Assert.False(b.IsEnabled(LogLevel.FromOrdinal(i)));
```
Fixed

304NotModified commented on tests/NLog.UnitTests/Config/RuleConfigurationTests.cs:
```diff
@@ -589,5 +591,221 @@ public void LoggingRule_LevelOff_NotSetAsActualLogLevel()
+            }
+        }
+
+        public static IEnumerable<object[]> LoggingRule_MinMaxLayout_ParseLevel_TestCases()
```
The idea by the [MemberData] was to reduce code duplication and make it easier to maintain. It seems we lost the latter one...

snakefoot replied:
You are welcome to unroll all the loops and generate one giant list.

304NotModified replied:
Is there a commit which I could use? Otherwise this will take some time.

snakefoot replied:
Sorry don't have a commit ready in my pocket :)

304NotModified replied:
I see these are now 80 test cases. I think this is hard to maintain and unneeded.  If we introduce a bug, it would be nice if 1 test fails, and not multiple. 

As rule of thumb I try to create these test cases:

- happy path
- alternatives paths 
- corner cases
- cases if needed for coverage.

That means no combination of cases, if it isn't needed for the list above.

For example this test: (shorter test, easier to show)

```c#
      [Theory]
        [InlineData("Off")]
        [InlineData("")]
        [InlineData((string)null)]
        [InlineData("Trace")]
        [InlineData("Debug")]
        [InlineData("Info")]
        [InlineData("Warn")]
        [InlineData("Error")]
        [InlineData(" error")]
        [InlineData("Fatal")]
        [InlineData("Wrong")]
        public void LoggingRule_LevelLayout_ParseLevel(string levelVariable
```
I think the cases "Debug", "Info", "Warn", "Error" (not " error") and "Fatal" not needed (redundant)

snakefoot replied:
You are right about not needing all the official Levels. I just wanted to make absolutely sure I didn't have an issue with min/max-levels crossing.

304NotModified commented on src/NLog/Config/LoggingRuleLevelFilter/DynamicLogLevelFilter.cs:
```diff
@@ -0,0 +1,132 @@
+            return new LoggingRuleLevelFilter(LogLevels);
+        }
+
+        private bool[] GenerateLogLevels()
```
how critical is the performance of this method?

snakefoot replied:
What are you guiding towards?

304NotModified replied:
refactoring

304NotModified replied:
by e.g replacing bool[] with a specialized class

snakefoot replied:
I can make refactoring for you without anyone being hurt.

snakefoot replied:
Refactoring complete

304NotModified replied:
If I would create a class that encapsulates bool[], do you think that would be a bad thing for the performance?

snakefoot replied:
No hidding the bool[] array away and making methods instead on `ILoggingRuleLevelFilter` should not hurt performance.

snakefoot replied:
But please make that in a new PR instead of adding more into this one.

304NotModified replied:
Ok, before of after merge this PR?

snakefoot replied:
After merge is easier for me :)

snakefoot commented:
@304NotModified Is the merge of this PR blocked by something?

304NotModified commented:
Same as the other PR, my free time at least.

Checked it in short, there are some things I like to get improved:

- LoggingRule_MinMaxLayout_ParseLevel_TestCases is really hard to maintain or even to reason which test cases this will generate. (maybe I have to check in VS2017, not done yet)
- There is some duplication in the testcases. As far as I remember, I tried to remove it, but it's still there/back?
- The ` LogLevels = new bool[LogLevel.MaxLevel.Ordinal + 1];` looks fragile and I prefer to encapsulate that in another object. I tried to help with some PRs, but have to conclude that commits of my on your branch aren't really helping - so I stopped with that for now.

304NotModified commented:
> Same as the other PR, my free time at least.
> 
> Checked it in short, there are some things I like to get improved:
> 
> * LoggingRule_MinMaxLayout_ParseLevel_TestCases is really hard to maintain or even to reason which test cases this will generate. (maybe I have to check in VS2017, not done yet)
> * There is some duplication in the testcases. As far as I remember, I tried to remove it, but it's still there/back?
> * The ` LogLevels = new bool[LogLevel.MaxLevel.Ordinal + 1];` looks fragile and I prefer to encapsulate that in another object. I tried to help with some PRs, but have to conclude that commits of my on your branch aren't really helping - so I stopped with that for now.

What's your opinion in this? This is still todo, isn't?

snakefoot commented:
> @304NotModified  What's your opinion in this? This is still todo, isn't?

Sounded like refactorings that that you wanted to implement later. And since not part of the public interface, then it is not blocking the merge of this feature.

304NotModified commented:
> Sounded like refactorings that that you wanted to implement later. And since not part of the public interface, then it is not blocking the merge of this feature.

Sorry, It's is blocking for me. Too much tech debt in this one.

snakefoot commented:
> Sorry, It's is blocking for me. Too much tech debt in this one.

So if I remove the LogLevel-array and replace with a class that has a property for each LogLevel, then you will be happy?

304NotModified commented:
i'm happy if the test cases aren't dynamic created and preferable with keeping things simple and reduce code duplication (I know that the latter two are a trade off)

snakefoot commented:
> i'm happy if the test cases aren't dynamic created and preferable with keeping things simple and reduce code duplication

So you will be happy if I just reduce the test to a few success-tests, and remove all test-combination? (Leaving LogLevel-array as it is).

304NotModified commented:
yes, there is no need to test every combination (it's even considered bad style)

snakefoot commented:
> yes, there is no need to test every combination (it's even considered bad style)

I just found this particular part of NLog very critical, and wanted to ensure it never broke. But will remove the yield-logic and just test a few good samples.

snakefoot commented:
@304NotModified Have now simplified the input-generation for the unit-tests.

Also changed target-branch to master, so it can be included in NLog 4.6.7

304NotModified commented:
What's changed in the force push? (Can't find the difference)

snakefoot commented:
> What's changed in the force push? (Can't find the difference)

No change beside pull + rebase,. Github was complaining it could not be merged right away.

304NotModified commented:
Ah , ok. You could ignore that warning. 

 I enabled that so it's clear before the merge it's targets the master ;)

304NotModified commented:
Merged :)

304NotModified commented:
Thanks!

snakefoot commented:
Tried to update the wiki: https://github.com/NLog/NLog/wiki/Filtering-log-messages
